### ISSUE TITLE
replication from compound using python

### ISSUE THREAD

**Issue by ciklysta** [irods/irods_rule_engine_plugin_python#103]
## Bug Report

Irods version 4.2.11, centos7

I have the following resource hierarchy (_OldResource_ being an old resource that I want to migrate data from)

```
DiskResource:unixfilesystem
OldResource:compound
├── OldArchiveResource:univmss
└── OldCacheResource:unixfilesystem
```

Data is only in the archive, the cache is empty.

I've written a custom rule that manages replication since it is a lengthy operation that I need to manage myself

_migrateOneObj.r_ content:
```
main {
    testReplicationFromCompound();
}
INPUT null
OUTPUT ruleExecOut
```

_core.re_ content:
```
testReplicationFromCompound {
    msiDataObjRepl("/zone/file.zip",
        "rescName=OldResource++++backupRescName=DiskResource", *Status);
}
```

When I run `irule -F migrateOneObj.r` under the user that owns `file.zip`, it works correctly.

However, if I move the function `testReplicationFromCompound` to python (_core.py_):
```python
def testReplicationFromCompound(rule_args, callback, rei):
    callback.msiDataObjRepl("/zone/file.zip",
        "rescName=OldResource++++backupRescName=DiskResource", 0)
```
and run `irule -F migrateOneObj.r` (under the user that owns `file.zip`) the following happens:
- the operation blocks on retrieving data from archive and never finishes
- there are 2 `irodsServer` processes, that started when `irule` was started. `strace` says that
  - first is blocked on `read` from a pipe
  - second (child of the previous one) is blocked on `futex(0x7f85e89d8ec8, FUTEX_WAIT_PRIVATE, 2, NULL`
- _rodsLog_'s last line is
  ```
  Jan 17 12:52:03 pid:88 NOTICE: execCmd:../../var/lib/irods/msiExecCmd_bin/migration- 
  interface.sh argv:stageToCache '/data/archive/dev/file.zip' '/data/cache/dev/file.zip
  ```
- univmss driver (shell script `migration-interface.sh`) is never run (otherwise it would create an entry in a custom log file)
- there is an empty file created in the cache resource vault
- in the sql db there are 2 rows:
  - first one corresponds to the _archive_ replica (data_is_dirty = 4)
  - second one corresponds to the _cache_ replica (data_is_dirty = 2)

**Comment by dworkin** [comment]
@trel @ciklysta This may be a duplicate of https://github.com/irods/irods/issues/6622 instead of #54: the problem occurs when `numThreads=1` and `msiExecCmd` is used with a python rule on the call stack.

**Comment by trel** [comment]
Oh, interesting...

Any chance you think that https://github.com/irods/irods/issues/6622 is actually, itself, the same as #54?

In other words, should we now re-test #54 to see if it is still a deadlock with the new https://github.com/irods/irods/issues/6622 codefix in place?

**Comment by dworkin** [comment]
Both issues deadlock on the same lock, both require a python rule on the call stack, but they are distinct.  #54 deadlocks without using `msiExecCmd`.

**Comment by trel** [comment]
Got it - right.
